docs/modules/architecture/pages/compact-binary-specification.adoc

```diff
@@ -0,0 +1,706 @@
+----
+
+The schema id is calculated from the byte array representation of the schema described above.
+
+The implementation that is used is as follows:
+
+[source,java,linenums]
+----
+long fingerprint64(byte[] buf) {
+  if (FP_TABLE == null) initFPTable();
+  long fp = EMPTY;
+  for (int i = 0; i < buf.length; i++)
+    fp = (fp >>> 8) ^ FP_TABLE[(int)(fp ^ buf[i]) & 0xff];
+  return fp;
+}
+
+static long EMPTY = 0xc15d213aa4d7a795L;
+static long[] FP_TABLE = null;
+
+void initFPTable() {
+  FP_TABLE = new long[256];
+  for (int i = 0; i < 256; i++) {
+    long fp = i;
+    for (int j = 0; j < 8; j++)
+      fp = (fp >>> 1) ^ (EMPTY & -(fp & 1L));
+    FP_TABLE[i] = fp;
+  }
+}
+----
+
+== Arrays
+
+Arrays of fix-sized items can not have `null` items. On the other hand, arrays of variable-size items may contain `null` items.
+
+=== Array of Fixed-size Items
+
+[cols=",",options="header",]
+|===
+|Name |Type
+|Number of items |i32
+|item 0 |item type
+|item 1 |item type
+|item 2 |item type
+|item n |item type
+|===
+
+=== Array of Variable-size Items
+
+Consists of `Header`, `Data`, and `Offsets` sections in this order.
+
+=== Header Section
+
+[cols="1,1"]
+|===
+|Name |Type
+|Data length |i32
+|Number of items |i32
+|===
+
+=== Data Section
+
+[cols="1,1"]
+|===
+|Name |Type
+|Item 0 | item type
+|Item 1 | item type
+|... | ...
+|Item n | item type
+|===
+
+=== Offsets Section
+
+[cols="1,1"]
+|===
+|Name |Type
+|Item 0 offset | u8/u16/i32
+|Item 1 offset | u8/u16/i32
+|... | ...
+|Item n offset | u8/u16/i32
+|===
+
+
+---
+
+An array can contain only a single type of item.
+In the case of `Compact[]` all the items must have the same schema, i.e their schema id must be equal. 
+
+Offsets are calculated from the beginning of the `DATA SECTION` shown in the table above.
+
+`Data Length` is the length of the `DATA SECTION` shown in the table above.
+
+Offset sizes vary depending on the Data Length.
+
+* Data Length <= `254`, offsets are `u8` (`255` is reserved for `null`)
+* Data Length <= `65534`, offsets are `u16` (`65535` is reserved for `null`)
+* Otherwise, offsets are `i32`.
+
+Items can be `null`. The corresponding offset will be set to `-1` in that case.
+
+== Nullable Values
+
+Fixed-size fields will always be on the binary and take up space. On the other hand, when variable-size fields are set to `null`, their offset will be set to `-1` in the binary, and no further data will be written.
```

Comment:
Fixed-size fields will be always on the binary and take up space. On the other hand, when variable-size fields are set to `null`, their offset will be set to `-1` in the binary, and no further data will be written.